lib/web/desktop.go

```diff
@@ -103,17 +103,41 @@ func createDesktopConnection(
 	//
 	// In the future, we may want to do something smarter like latency-based
 	// routing.
-	service := winServices[rand.Intn(len(winServices))]
-	log = log.WithField("windows-service-uuid", service.GetName())
-	log = log.WithField("windows-service-addr", service.GetAddr())
-	serviceCon, err := site.DialTCP(reversetunnel.DialParams{
-		From:     &utils.NetAddr{AddrNetwork: "tcp", Addr: r.RemoteAddr},
-		To:       &utils.NetAddr{AddrNetwork: "tcp", Addr: service.GetAddr()},
-		ConnType: types.WindowsDesktopTunnel,
-		ServerID: service.GetName() + "." + ctx.parent.clusterName,
+	rand.Shuffle(len(winServices), func(i, j int) {
```

Comment:
You're still fetching _all_ Windows Desktop Services, then shuffling them, and then searching through them until you find one that works for this desktop.

If there are thousands of Windows Desktop Services and only 2-3 of them are actually eligible candidates, this is wasteful.

What if you did something like this (pseudocode) instead?

```
desktops, _ := GetWindowsDesktopsByName(desktopName)

var eligibleServiceIDs []string
for _, desktop := range desktops {
    eligibleServiceIDs = append(eligibleServiceIDs, desktop.GetHostID())
}
rand.Shuffle(eligibleServiceIDs, ...)

// try each eligible service, one at a time
// (this will likely be a much smaller list than your original for loop)
for _, id := eligibleServiceIDs {
    wds, _ := GetWindowsDesktopService(id) // this would be a new RPC
    serviceCon, err = site.DialTCP(...)
}
```